fw/apm.c

```diff
@@ -1157,83 +1173,140 @@ tfw_apm_create(void)
 		TfwApmUBuf *ubuf = per_cpu_ptr(data->ubuf, icpu);
 		ubent = kmalloc_node(size, GFP_KERNEL, cpu_to_node(icpu));
 		if (!ubent)
-			goto cleanup;
+			return ERR_PTR(-ENOMEM);
 		for (i = 0; i < 2 * TFW_APM_UBUF_SZ; ++i)
 			WRITE_ONCE(ubent[i].data, ULONG_MAX);
 		ubuf->ubent[0] = ubent;
 		ubuf->ubent[1] = ubent + TFW_APM_UBUF_SZ;
 		ubuf->ubufsz = TFW_APM_UBUF_SZ;
 	}
 
+	/* Return end of the structure, for further memory areas setting. */
+	return val[1] + T_PSZ;
+}
+
+static TfwApmRef *
+tfw_apm_ref_create(void)
+{
+	int i;
+	int rbufsz = tfw_apm_tmwscale;
+	int hm_size = tfw_hm_codes_cnt * sizeof(TfwApmHMStats);
+	int size = sizeof(TfwApmRef)
+		+ rbufsz * sizeof(TfwApmRBEnt)
+		+ 2 * T_PSZ * sizeof(unsigned int)
+		+ hm_size;
+	TfwApmRef *ref;
+	TfwApmHMStats *hmstats;
+	TfwApmHMCfg *ent;
+
+	if ((ref = kzalloc(size, GFP_ATOMIC)) == NULL)
+		return ERR_PTR(-ENOMEM);
+
+	/*
+	 * End of the TfwApmData structure is the beginning of the TfwApmHMStats
+	 * of the TfwApmHMCtl.
+	 */
+	hmstats = tfw_apm_data_init(&ref->data);
```

Comment:
It looks really weird getting the pointer to the end of the initialized data. This looks like an obfuscation. The original variant was more logical and correct, you have all data here to calculate the start of the area for stats.

Reply:
Actually, it doesn't. `tfw_apm_data_init()` initializes the `TfwApmData{}` structure. The whole layout of the structure is defined here, and the memory for that layout is allocated here as well. So, whatever implementation it might be, it must follow the defined layout, and it doesn't go beyond that. Nor should it.
The layout defined here includes the memory for `TfwApmHMStats{}` at the end. You don't need the pointer to that area from `tfw_apm_data_init()`, that's very obscure. You have all data here to calculate where that area of memory is located. It's `ref + sizeof(TfwApmRef) + rbufsz * sizeof(TfwApmRBEnt) + 2 * T_PSZ * sizeof(unsigned int)` - which can be calculated once at the start.